python/taichi/lang/kernel_impl.py

```diff
@@ -486,11 +490,22 @@ def materialize(self, key=None, args=None, arg_features=None):
         if key is None:
             key = (self.func, 0)
         self.runtime.materialize()
+
+        # Transform the primal kernel to forward mode grad kernel
+        # then recover to primal when exiting the forward mode manager
+        if self.runtime.fwd_mode_manager:
+            self.autodiff_mode = AutodiffMode.FORWARD
+            self.runtime.fwd_mode_manager.insert(self)
+            self.compiled_functions = self.runtime.compiled_fwd_mode_grad_functions
```

Comment:
why do we need to replace compiled_functions?

Reply:
The reason is to distinguish the primal and forward grad kernel.
The two kernels has the same name and args, thus the func key is the same, i.e., one will overwrite another in a map.